### PR TITLE
test: use MiniWallet for rpc_createmultisig.py

### PR DESCRIPTION
**Pull request by ayush933**
This PR enables one of the non-wallet functional tests (rpc_createmultisig.py) to be run even with the Bitcoin Core wallet disabled by using the MiniWallet instead, as proposed in #20078 .

### CODE REVIEW

File: test/functional/rpc_createmultisig.py [hunk 1]
```diff
@@ -18,15 +18,23 @@
     assert_equal,
 )
 from test_framework.wallet_util import bytes_to_wif
+from test_framework.wallet import (
+    MiniWallet,
+    address_to_scriptpubkey,
+    getnewdestination
```
Comment (MarcoFalke) [inline test/functional/rpc_createmultisig.py:24]: nit: add trailing comma
```suggestion
    getnewdestination,
```

File: test/functional/rpc_createmultisig.py [hunk 2]
```diff
@@ -18,15 +18,23 @@
...
         self.setup_clean_chain = True
         self.num_nodes = 3
         self.supports_cli = False
+        if self.is_wallet_compiled():
+            self.requires_wallet = True
 
     def skip_test_if_missing_module(self):
-        self.skip_if_no_wallet()
+        if self.is_wallet_compiled() and not self.options.descriptors:
+            self.skip_if_no_bdb()
```
Comment (MarcoFalke) [inline test/functional/rpc_createmultisig.py:37]: Is this check still needed if you adjust the checks below to check for bdb instead?

**Review by maflcko — APPROVED** [review]
Thanks, Concept ACK. Some questions.

File: test/functional/rpc_createmultisig.py (hunk 1 above)
Comment (ayush933) [reply test/functional/rpc_createmultisig.py]: done

File: test/functional/rpc_createmultisig.py (hunk 2 above)
Comment (ayush933) [reply test/functional/rpc_createmultisig.py]: Changed the checks to check for bdb instead. Removed this check.

File: test/functional/rpc_createmultisig.py [hunk 3]
```diff
@@ -37,57 +45,61 @@ def get_keys(self):
...
             self.priv.append(bytes_to_wif(k.get_bytes(), k.is_compressed))
-        self.final = node2.getnewaddress()
+        self.final = getnewdestination()[2]
 
     def run_test(self):
         node0, node1, node2 = self.nodes
+        self.wallet=MiniWallet(test_node=node0)
 
-        self.check_addmultisigaddress_errors()
+        if self.is_wallet_compiled():
```
[the remark this answers is not in the document]
Comment (ayush933) [reply test/functional/rpc_createmultisig.py]: Changed to check for bdb instead.

File: test/functional/rpc_createmultisig.py [hunk 4]
```diff
@@ -18,15 +18,19 @@
...
 class RpcCreateMultiSigTest(BitcoinTestFramework):
     def set_test_params(self):
         self.setup_clean_chain = True
         self.num_nodes = 3
         self.supports_cli = False
-
-    def skip_test_if_missing_module(self):
-        self.skip_if_no_wallet()
+        if self.is_bdb_compiled():
+            self.requires_wallet = True
```
Comment (MarcoFalke) [inline test/functional/rpc_createmultisig.py:33]: Why is this needed? IIRC this will only import the coinbase keys, but they aren't needed?

File: test/functional/rpc_createmultisig.py [hunk 5]
```diff
@@ -164,17 +160,23 @@ def do_multisig(self):
...
+            maddw = msigw["address"]
+            mredeemw = msigw["redeemScript"]
+            assert_equal(desc, drop_origins(msigw['descriptor']))
+            # addmultisigiaddress and createmultisig work the same
+            assert maddw == madd
+            assert mredeemw == mredeem
+            wmulti.unloadwallet()
+
+        if self.output_type == "bech32":
+            spk = bytes.fromhex(node0.validateaddress(madd)["scriptPubKey"])
```
Comment (MarcoFalke) [inline test/functional/rpc_createmultisig.py:175]: I think you can remove the if-else and just unconditionally use `validateaddress` for simplicity.

File: test/functional/rpc_createmultisig.py [hunk 6]
```diff
@@ -119,33 +127,21 @@ def check_addmultisigaddress_errors(self):
...
-        bal0 = node0.getbalance()
-        bal1 = node1.getbalance()
-        bal2 = node2.getbalance()
-
-        height = node0.getblockchaininfo()["blocks"]
-        assert 150 < height < 350
-        total = 149 * 50 + (height - 149 - 100) * 25
-        assert bal1 == 0
-        assert bal2 == self.moved
-        assert bal0 + bal1 + bal2 == total
```
Comment (MarcoFalke) [inline test/functional/rpc_createmultisig.py:135]: Is there a reason why this check fails? IIUC the multisig addresses are still added to the bdb wallets as watch-only, so the watchonly balance should still match up (considering also the MiniWallet balance)

File: test/functional/rpc_createmultisig.py (hunk 4 above)
Comment (ayush933) [reply test/functional/rpc_createmultisig.py:33]: This is actually needed to create default wallets. Without this `check_addmultisigaddress_errors()` fails.

File: test/functional/rpc_createmultisig.py (hunk 6 above)
Comment (ayush933) [reply test/functional/rpc_createmultisig.py:135]: Sorry but I have been stuck here for quite some time and not able to figure out why this fails.

Comment (MarcoFalke) [reply test/functional/rpc_createmultisig.py:135]: Ok, I can take a look later.

**Comment by brunoerg** [comment]
Concept ACK.

You should squash the commits, see: https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

File: test/functional/rpc_createmultisig.py (hunk 6 above)
Comment (MarcoFalke) [reply test/functional/rpc_createmultisig.py:135]: This works for me locally:


```diff
diff --git a/test/functional/rpc_createmultisig.py b/test/functional/rpc_createmultisig.py
index 4cf49fbcd3..9106fb1366 100755
--- a/test/functional/rpc_createmultisig.py
+++ b/test/functional/rpc_createmultisig.py
@@ -39,7 +39,10 @@ class RpcCreateMultiSigTest(BitcoinTestFramework):
             k.generate()
             self.pub.append(k.get_pubkey().get_bytes().hex())
             self.priv.append(bytes_to_wif(k.get_bytes(), k.is_compressed))
-        self.final = getnewdestination()[2]
+        if self.is_bdb_compiled():
+         self.final = node2.getnewaddress()
+        else:
+         self.final = getnewdestination()[2]
 
     def run_test(self):
         node0, node1, node2 = self.nodes
@@ -51,11 +54,13 @@ class RpcCreateMultiSigTest(BitcoinTestFramework):
         self.log.info('Generating blocks ...')
         self.generate(self.wallet, 149)
 
+        self.moved=0
         for self.nkeys in [3, 5]:
             for self.nsigs in [2, 3]:
                 for self.output_type in ["bech32", "p2sh-segwit", "legacy"]:
                     self.get_keys()
                     self.do_multisig()
+        self.checkbalances()
 
 
         # Test mixed compressed and uncompressed pubkeys
@@ -125,6 +130,21 @@ class RpcCreateMultiSigTest(BitcoinTestFramework):
         pubs = [self.nodes[1].getaddressinfo(addr)["pubkey"] for addr in addresses]
         assert_raises_rpc_error(-5, "Bech32m multisig addresses cannot be created with legacy wallets", self.nodes[0].addmultisigaddress, 2, pubs, "", "bech32m")
 
+    def checkbalances(self):
+        node0, node1, node2 = self.nodes
+        self.generate(node0, 100)#COINBASE_MATURITY)
+
+        bal0 = node0.getbalance()
+        bal1 = node1.getbalance()
+        bal2 = node2.getbalance()
+        balw=self.wallet.get_balance()
+
+        height = node0.getblockchaininfo()["blocks"]
+        assert 150 < height < 350
+        total = 149 * 50 + (height - 149 - 100) * 25
+        assert bal1 == 0
+        assert bal2 == self.moved
+        assert_equal (bal0 + bal1 + bal2 + balw ,total) 
 
     def do_multisig(self):
         node0, node1, node2 = self.nodes
@@ -213,6 +233,7 @@ class RpcCreateMultiSigTest(BitcoinTestFramework):
         rawtx2 = node2.signrawtransactionwithkey(rawtx, self.priv[0:self.nsigs - 1], prevtxs)
         rawtx3 = node2.signrawtransactionwithkey(rawtx2["hex"], [self.priv[-1]], prevtxs)
 
+        self.moved += outval
         tx = node0.sendrawtransaction(rawtx3["hex"], 0)
         blk = self.generate(node0, 1)[0]
         assert tx in node0.getblock(blk)["tx"]
diff --git a/test/functional/test_framework/wallet.py b/test/functional/test_framework/wallet.py
index dd41a740ae..979858e2ee 100644
--- a/test/functional/test_framework/wallet.py
+++ b/test/functional/test_framework/wallet.py
@@ -93,6 +93,9 @@ class MiniWallet:
             self._address, self._internal_key = create_deterministic_address_bcrt1_p2tr_op_true()
             self._scriptPubKey = bytes.fromhex(self._test_node.validateaddress(self._address)['scriptPubKey'])
 
+    def get_balance(self):
+        return sum(u['value'] for u in self._utxos)
+
     def rescan_utxos(self):
         """Drop all utxos and rescan the utxo set"""
         self._utxos = []

Comment (ayush933) [reply test/functional/rpc_createmultisig.py:135]: Made the necessary changes and squashed commits.

**Review by maflcko — APPROVED** [review]

**Comment by theStack** [comment]
Concept ACK

**Review by danielabrozzoni — APPROVED** [review]
tACK 4ea7c1b410bc6626a14c2cac8841910bc51e2422 - the code looks good, the test works as intended even when the wallet is disabled.

File: test/functional/test_runner.py [hunk 7]
```diff
@@ -225,8 +225,7 @@
     'feature_rbf.py --descriptors',
     'mempool_packages.py',
     'mempool_package_onemore.py',
-    'rpc_createmultisig.py --legacy-wallet',
```
Comment (michaelfolkson) [inline test/functional/test_runner.py:228]: Unless I'm mistaken you do want both tests run (--legacy-wallet and --descriptors). We are still supporting the legacy wallet and hence we should still test it. I think you've taken it out because of the discussion in #24635 but the legacy wallet is [not yet deprecated](https://github.com/bitcoin/bitcoin/issues/20160).

**Review by michaelfolkson — COMMENTED** [review]
Concept ACK

This is one of the tests from #20078 that could use MiniWallet instead of being skipped when the Core wallet is disabled.

File: test/functional/test_runner.py (hunk 7 above)
Comment (ayush933) [reply test/functional/test_runner.py:228]: In this particular test, the wallet, when compiled is used for `addmultisigaddress` which descriptor wallet doesn't actually support, so I don't think there is any difference in running with or without the flag.
The legacy wallet when compiled is actually used even without the flag so there is no need to run the test twice.

**Review by maflcko — APPROVED** [review]

**Comment by fanquake** [comment]
@danielabrozzoni / @theStack want to re-ACK?

**Comment by michaelfolkson** [comment]
I'll test this and ACK it in the next couple of days.

**Comment by danielabrozzoni** [comment]
re-ACK 2726b60a3ac098b44f2970bed21147b70e12a1c2
